pkg/userop/client.go

```diff
@@ -0,0 +1,454 @@
+	switch *c.smartWallet.Type {
+	case SmartWalletSimpleAccount, SmartWalletBiconomy:
+		return handleCallSimpleAccount(calls)
+	case SmartWalletKernel:
+		return handleCallKernel(calls)
+	default:
+		return nil, fmt.Errorf("unknown smart wallet type: %s", c.smartWallet.Type)
+	}
+}
+
+// handleCallSimpleAccount packs CallData for SimpleAccount smart wallet.
+func handleCallSimpleAccount(calls []Call) ([]byte, error) {
+	addresses := make([]any, 0, len(calls))
+	calldatas := make([]any, 0, len(calls))
+	for _, call := range calls {
+		addresses = append(addresses, call.To)
+		calldatas = append(calldatas, call.CallData)
+	}
+
+	// Pack the data for the `executeBatch` smart account function.
+	// Biconomy v2.0: https://github.com/bcnmy/scw-contracts/blob/v2-deployments/contracts/smart-account/SmartAccount.sol#L128
+	// NOTE: you can NOT send native token with SimpleAccount v0.6.0 because of `executeBatch` signature
+	data, err := simpleAccountABI.Pack("executeBatch", addresses, calldatas)
+	if err != nil {
+		return nil, fmt.Errorf("failed to pack executeBatch data for SimpleAccount: %w", err)
+	}
+	return data, nil
+}
+
+// callStructKernel represents a call to the Zerodev Kernel smart wallet.
+// The idea is the same as in Call type,
+// but tailed specifically to the Zerodev Kernel ABI.
+type callStructKernel struct {
+	To    common.Address `json:"to"`
+	Value *big.Int       `json:"value"`
+	Data  []byte         `json:"data"`
+}
+
+// handleCallKernel packs calldata for Zerodev Kernel smart wallet.
+func handleCallKernel(calls []Call) ([]byte, error) {
+	params := make([]callStructKernel, 0, len(calls))
+	for _, call := range calls {
+		params = append(params, callStructKernel{
+			To:    call.To,
+			Value: call.Value.BigInt(),
+			Data:  call.CallData,
+		})
+	}
+
+	// pack the data for the `executeBatch` smart account function
+	// Zerodev Kernel v2.2: https://github.com/zerodevapp/kernel/blob/807b75a4da6fea6311a3573bc8b8964a34074d94/src/Kernel.sol#L82
+	data, err := kernelExecuteABI.Pack("executeBatch", params)
+	if err != nil {
+		return nil, fmt.Errorf("failed to pack executeBatch data for Kernel: %w", err)
+	}
+	return data, nil
+}
+
+// waitForUserOpEvent waits for a user operation to be committed on block.
+func waitForUserOpEvent(
+	ctx context.Context,
+	cancel context.CancelFunc,
+	client EthBackend,
+	done chan<- Receipt,
+	entryPoint common.Address,
+	userOpHash common.Hash,
+) {
+	ticker := time.NewTicker(time.Millisecond * 5000)
+	defer ticker.Stop()
+	defer close(done)
+
+	fromBlock, err := client.BlockNumber(ctx)
+	if err != nil {
+		slog.Error("failed to get block number", "error", err)
+		cancel()
+		return
+	}
+
+	query := ethereum.FilterQuery{
+		Addresses: []common.Address{entryPoint},
+		Topics:    [][]common.Hash{{}, {userOpHash}},
+		FromBlock: big.NewInt(int64(fromBlock)),
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			slog.Error("timeout waiting for user operation event", "hash", userOpHash.Hex())
+			return
```

Comment:
Maybe you should close the done channel here?

Reply:
It's closed in `defer close(done)` on line 393